```
FAILED scipy/fft/tests/test_basic.py::TestFFT1D::test_rfft[numpy] - AttributeError: module 'numpy' has no attribute 'astype'
```
that is why I avoided it in test code I think. In the library code, I think we always ensure to use the compat library, but not in tests.

--- a/.github/workflows/array_api.yml
+++ b/.github/workflows/array_api.yml
@@ -58,7 +58,7 @@ jobs:
 
     - name: Install Python packages
       run: |
-        python -m pip install numpy cython pytest pytest-xdist pytest-timeout pybind11 mpmath gmpy2 pythran ninja meson click rich-click doit pydevtool pooch hypothesis
+        python -m pip install numpy cython pytest pytest-xdist pytest-timeout pybind11 mpmath gmpy2 pythran ninja meson click rich-click doit pydevtool pooch hypothesis array-api-strict
 
     - name: Install PyTorch CPU
       run: |
@@ -92,6 +92,6 @@ jobs:
         export SCIPY_USE_PROPACK=1
         # expand as new modules are added
         # --array-api-backend == -b
-        python dev.py --no-build test -b pytorch -b numpy -b numpy.array_api -s cluster -- --durations 3 --timeout=60
-        python dev.py --no-build test -b pytorch -b numpy -b numpy.array_api -s fft -- --durations 3 --timeout=60
+        python dev.py --no-build test -b pytorch -b numpy -b array_api_strict -s cluster -- --durations 3 --timeout=60
+        python dev.py --no-build test -b pytorch -b numpy -b array_api_strict -s fft -- --durations 3 --timeout=60
         python dev.py --no-build test --array-api-backend all --tests scipy/_lib/tests/test_array_api.py

--- a/dev.py
+++ b/dev.py
@@ -689,7 +689,7 @@ class Test(Task):
         ['--array-api-backend', '-b'], default=None, metavar='ARRAY_BACKEND',
         multiple=True,
         help=(
-            "Array API backend ('all', 'numpy', 'pytorch', 'cupy', 'numpy.array_api')."
+            "Array API backend ('all', 'numpy', 'pytorch', 'cupy', 'array_api_strict')."
         )
     )
     # Argument can't have `help=`; used to consume all of `-- arg1 arg2 arg3`

--- a/environment.yml
+++ b/environment.yml
@@ -29,6 +29,7 @@ dependencies:
   - pytest-timeout
   - asv >=0.6
   - hypothesis
+  - array-api-strict
   # For type annotations
   - mypy
   - typing_extensions

--- a/mypy.ini
+++ b/mypy.ini
@@ -79,6 +79,9 @@ ignore_missing_imports = True
 [mypy-appdirs]
 ignore_missing_imports = True
 
+[mypy-array_api_strict]
+ignore_missing_imports = True
+
 #
 # Extension modules without stubs.
 #

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -91,6 +91,7 @@ test = [
     "scikit-umfpack",
     "pooch",
     "hypothesis>=6.30",
+    "array-api-strict",
 ]
 doc = [
     "sphinx>=5.0.0",

--- a/pytest.ini
+++ b/pytest.ini
@@ -14,6 +14,5 @@ filterwarnings =
     ignore:'contextfunction' is renamed to 'pass_context'
     ignore:.*The distutils.* is deprecated.*:DeprecationWarning
     ignore:\s*.*numpy.distutils.*:DeprecationWarning
-    ignore:.*The numpy.array_api submodule is still experimental.*:UserWarning
     ignore:.*`numpy.core` has been made officially private.*:DeprecationWarning
     ignore:.*In the future `np.long` will be defined as.*:FutureWarning

--- a/requirements/test.txt
+++ b/requirements/test.txt
@@ -11,3 +11,4 @@ threadpoolctl
 # scikit-umfpack  # circular dependency issues
 pooch
 hypothesis>=6.30
+array-api-strict

--- a/scipy/cluster/hierarchy.py
+++ b/scipy/cluster/hierarchy.py
@@ -161,7 +161,7 @@ def _warning(s):
 
 
 def int_floor(arr, xp):
-    # numpy.array_api is strict about not allowing `int()` on a float array.
+    # array_api_strict is strict about not allowing `int()` on a float array.
     # That's typically not needed, here it is - so explicitly convert
     return int(xp.astype(xp.asarray(arr), xp.int64))
 

--- a/scipy/cluster/tests/test_hierarchy.py
+++ b/scipy/cluster/tests/test_hierarchy.py
@@ -1271,7 +1271,7 @@ def test_node_compare(xp):
 
 @skip_if_array_api_gpu
 @array_api_compatible
-@skip_if_array_api_backend('numpy.array_api')
+@skip_if_array_api_backend('array_api_strict')
 def test_cut_tree(xp):
     np.random.seed(23)
     nobs = 50

--- a/scipy/conftest.py
+++ b/scipy/conftest.py
@@ -6,7 +6,6 @@ import tempfile
 from functools import wraps
 
 import numpy as np
-import numpy.array_api
 import numpy.testing as npt
 import pytest
 import hypothesis
@@ -108,7 +107,11 @@ xp_available_backends = {'numpy': np}
 
 if SCIPY_ARRAY_API and isinstance(SCIPY_ARRAY_API, str):
     # fill the dict of backends with available libraries
-    xp_available_backends.update({'numpy.array_api': numpy.array_api})
+    try:
+        import array_api_strict
+        xp_available_backends.update({'array_api_strict': array_api_strict})
+    except ImportError:
+        pass
 
     try:
         import torch  # type: ignore[import]

--- a/scipy/fft/tests/test_basic.py
+++ b/scipy/fft/tests/test_basic.py
@@ -15,6 +15,25 @@ from scipy._lib._array_api import (
     array_namespace, size, xp_assert_close, xp_assert_equal
 )
 
+
+# Expected input dtypes. Note that `scipy.fft` is more flexible for numpy,
+# but for C2C transforms like `fft.fft`, the array API standard only mandates
+# that complex dtypes should work, float32/float64 aren't guaranteed to.
+def get_expected_input_dtype(func, xp):
+    if func in [fft.fft, fft.fftn, fft.fft2,
+                fft.ifft, fft.ifftn, fft.ifft2,
+                fft.hfft, fft.hfftn, fft.hfft2,
+                fft.irfft, fft.irfftn, fft.irfft2]:
+        dtype = xp.complex128
+    elif func in [fft.rfft, fft.rfftn, fft.rfft2,
+                  fft.ihfft, fft.ihfftn, fft.ihfft2]:
+        dtype = xp.float64
+    else:
+        raise ValueError(f'Unknown FFT function: {func}')
+
+    return dtype
+
+
 def fft1(x):
     L = len(x)
     phase = -2j*np.pi*(np.arange(L)/float(L))
@@ -26,7 +45,7 @@ class TestFFTShift:
 
     @array_api_compatible
     def test_fft_n(self, xp):
-        x = xp.asarray([1, 2, 3])
+        x = xp.asarray([1, 2, 3], dtype=xp.complex128)
         if xp.__name__ == 'torch':
             assert_raises(RuntimeError, fft.fft, x, 0)
         else:
@@ -114,11 +133,12 @@ class TestFFT1D:
 
     @array_api_compatible
     def test_rfft(self, xp):
-        x = xp.asarray(random(29))
+        x = xp.asarray(random(29), dtype=xp.float64)
         for n in [size(x), 2*size(x)]:
             for norm in [None, "backward", "ortho", "forward"]:
                 xp_assert_close(fft.rfft(x, n=n, norm=norm),
-                                fft.fft(x, n=n, norm=norm)[:(n//2 + 1)])
+                                fft.fft(xp.asarray(x, dtype=xp.complex128),
+                                        n=n, norm=norm)[:(n//2 + 1)])
             xp_assert_close(
                 fft.rfft(x, n=n, norm="ortho"),
                 fft.rfft(x, n=n) / xp.sqrt(xp.asarray(n, dtype=xp.float64))
@@ -135,8 +155,8 @@ class TestFFT1D:
     @skip_if_array_api_backend('torch')
     @array_api_compatible
     def test_rfft2(self, xp):
-        x = xp.asarray(random((30, 20)))
-        expect = fft.fft2(x)[:, :11]
+        x = xp.asarray(random((30, 20)), dtype=xp.float64)
+        expect = fft.fft2(xp.asarray(x, dtype=xp.complex128))[:, :11]
         xp_assert_close(fft.rfft2(x), expect)
         xp_assert_close(fft.rfft2(x, norm="backward"), expect)
         xp_assert_close(fft.rfft2(x, norm="ortho"),
@@ -156,8 +176,8 @@ class TestFFT1D:
     # torch.fft not yet implemented by array-api-compat
     @skip_if_array_api_backend('torch')
     def test_rfftn(self, xp):
-        x = xp.asarray(random((30, 20, 10)))
-        expect = fft.fftn(x)[:, :, :6]
+        x = xp.asarray(random((30, 20, 10)), dtype=xp.float64)
+        expect = fft.fftn(xp.asarray(x, dtype=xp.complex128))[:, :, :6]
         xp_assert_close(fft.rfftn(x), expect)
         xp_assert_close(fft.rfftn(x, norm="backward"), expect)
         xp_assert_close(fft.rfftn(x, norm="ortho"),
@@ -211,8 +231,8 @@ class TestFFT1D:
     @skip_if_array_api_backend('torch')
     @array_api_compatible
     def test_ihfft2(self, xp):
-        x = xp.asarray(random((30, 20)))
-        expect = fft.ifft2(x)[:, :11]
+        x = xp.asarray(random((30, 20)), dtype=xp.float64)
+        expect = fft.ifft2(xp.asarray(x, dtype=xp.complex128))[:, :11]
         xp_assert_close(fft.ihfft2(x), expect)
         xp_assert_close(fft.ihfft2(x, norm="backward"), expect)
         xp_assert_close(
@@ -234,8 +254,8 @@ class TestFFT1D:
     # torch.fft not yet implemented by array-api-compat
     @skip_if_array_api_backend('torch')
     def test_ihfftn(self, xp):
-        x = xp.asarray(random((30, 20, 10)))
-        expect = fft.ifftn(x)[:, :, :6]
+        x = xp.asarray(random((30, 20, 10)), dtype=xp.float64)
+        expect = fft.ifftn(xp.asarray(x, dtype=xp.complex128))[:, :, :6]
         xp_assert_close(expect, fft.ihfftn(x))
         xp_assert_close(expect, fft.ihfftn(x, norm="backward"))
         xp_assert_close(
@@ -245,7 +265,8 @@ class TestFFT1D:
         xp_assert_close(fft.ihfftn(x, norm="forward"), expect * (30 * 20 * 10))
 
     def _check_axes(self, op, xp):
-        x = xp.asarray(random((30, 20, 10)))
+        dtype = get_expected_input_dtype(op, xp)
+        x = xp.asarray(random((30, 20, 10)), dtype=dtype)
         axes = [(0, 1, 2), (0, 2, 1), (1, 0, 2), (1, 2, 0), (2, 0, 1), (2, 1, 0)]
         xp_test = array_namespace(x)
         for a in axes:
@@ -273,7 +294,8 @@ class TestFFT1D:
     @pytest.mark.parametrize("op", [fft.fftn, fft.ifftn,
                                     fft.rfftn, fft.irfftn])
     def test_axes_subset_with_shape_standard(self, op, xp):
-        x = xp.asarray(random((16, 8, 4)))
+        dtype = get_expected_input_dtype(op, xp)
+        x = xp.asarray(random((16, 8, 4)), dtype=dtype)
         axes = [(0, 1, 2), (0, 2, 1), (1, 2, 0)]
         xp_test = array_namespace(x)
         for a in axes:
@@ -295,7 +317,8 @@ class TestFFT1D:
                                     fft.hfft2, fft.ihfft2,
                                     fft.hfftn, fft.ihfftn])
     def test_axes_subset_with_shape_non_standard(self, op, xp):
-        x = xp.asarray(random((16, 8, 4)))
+        dtype = get_expected_input_dtype(op, xp)
+        x = xp.asarray(random((16, 8, 4)), dtype=dtype)
         axes = [(0, 1, 2), (0, 2, 1), (1, 2, 0)]
         xp_test = array_namespace(x)
         for a in axes:
@@ -310,17 +333,21 @@ class TestFFT1D:
     @array_api_compatible
     def test_all_1d_norm_preserving(self, xp):
         # verify that round-trip transforms are norm-preserving
-        x = xp.asarray(random(30))
+        x = xp.asarray(random(30), dtype=xp.float64)
         xp_test = array_namespace(x)
         x_norm = xp_test.linalg.vector_norm(x)
         n = size(x) * 2
-        func_pairs = [(fft.fft, fft.ifft),
-                      (fft.rfft, fft.irfft),
+        func_pairs = [(fft.rfft, fft.irfft),
                       # hfft: order so the first function takes x.size samples
                       #       (necessary for comparison to x_norm above)
                       (fft.ihfft, fft.hfft),
+                      # functions that expect complex dtypes at the end
+                      (fft.fft, fft.ifft),
                       ]
         for forw, back in func_pairs:
+            if forw == fft.fft:
+                x = xp.asarray(x, dtype=xp.complex128)
+                x_norm = xp_test.linalg.vector_norm(x)
             for n in [size(x), 2*size(x)]:
                 for norm in ['backward', 'ortho', 'forward']:
                     tmp = forw(x, n=n, norm=norm)
@@ -346,20 +373,25 @@ class TestFFT1D:
 
     @array_api_compatible
     @pytest.mark.parametrize("dtype", ["float32", "float64"])
-    def test_dtypes(self, dtype, xp):
+    def test_dtypes_real(self, dtype, xp):
         x = xp.asarray(random(30), dtype=getattr(xp, dtype))
-        out_dtypes = {"float32": xp.complex64, "float64": xp.complex128}
-        x_complex = xp.asarray(x, dtype=out_dtypes[dtype])
 
-        res_fft = fft.ifft(fft.fft(x))
         res_rfft = fft.irfft(fft.rfft(x))
         res_hfft = fft.hfft(fft.ihfft(x), x.shape[0])
         # Check both numerical results and exact dtype matches
         rtol = {"float32": 1.2e-4, "float64": 1e-8}[dtype]
-        xp_assert_close(res_fft, x_complex, rtol=rtol, atol=0)
         xp_assert_close(res_rfft, x, rtol=rtol, atol=0)
         xp_assert_close(res_hfft, x, rtol=rtol, atol=0)
 
+    @array_api_compatible
+    @pytest.mark.parametrize("dtype", ["complex64", "complex128"])
+    def test_dtypes_complex(self, dtype, xp):
+        x = xp.asarray(random(30), dtype=getattr(xp, dtype))
+
+        res_fft = fft.ifft(fft.fft(x))
+        # Check both numerical results and exact dtype matches
+        rtol = {"complex64": 1.2e-4, "complex128": 1e-8}[dtype]
+        xp_assert_close(res_fft, x, rtol=rtol, atol=0)
 
 @pytest.mark.parametrize(
         "dtype",
@@ -490,8 +522,13 @@ class TestIRFFTN:
                                   fft.fftn, fft.ifftn,
                                   fft.rfftn, fft.irfftn, fft.hfft, fft.ihfft])
 def test_non_standard_params(func, xp):
+    if func in [fft.rfft, fft.rfftn, fft.ihfft]:
+        dtype = xp.float64
+    else:
+        dtype = xp.complex128
+
     if xp.__name__ != 'numpy':
-        x = xp.asarray([1, 2, 3])
+        x = xp.asarray([1, 2, 3], dtype=dtype)
         # func(x) should not raise an exception
         func(x)
         assert_raises(ValueError, func, x, workers=2)

--- a/scipy/fft/tests/test_helper.py
+++ b/scipy/fft/tests/test_helper.py
@@ -426,8 +426,8 @@ class TestFFTShift:
 
 class TestFFTFreq:
 
-    # fft not yet implemented by numpy.array_api
-    @skip_if_array_api_backend('numpy.array_api')
+    # fft not yet implemented by array-api-strict
+    @skip_if_array_api_backend('array_api_strict')
     # cupy.fft not yet implemented by array-api-compat
     @skip_if_array_api_backend('cupy')
     @array_api_compatible
@@ -456,8 +456,8 @@ class TestFFTFreq:
 
 class TestRFFTFreq:
 
-    # fft not yet implemented by numpy.array_api
-    @skip_if_array_api_backend('numpy.array_api')
+    # fft not yet implemented by array-api-strict
+    @skip_if_array_api_backend('array_api_strict')
     # cupy.fft not yet implemented by array-api-compat
     @skip_if_array_api_backend('cupy')
     @array_api_compatible

--- a/scipy/fft/tests/test_helper.py
+++ b/scipy/fft/tests/test_helper.py
@@ -344,12 +344,12 @@ class TestFFTShift:
     @skip_if_array_api_backend('torch')
     @array_api_compatible
     def test_definition(self, xp):
-        x = xp.asarray([0, 1, 2, 3, 4, -4, -3, -2, -1])
-        y = xp.asarray([-4, -3, -2, -1, 0, 1, 2, 3, 4])
+        x = xp.asarray([0., 1, 2, 3, 4, -4, -3, -2, -1])
+        y = xp.asarray([-4., -3, -2, -1, 0, 1, 2, 3, 4])
         xp_assert_close(fft.fftshift(x), y)
         xp_assert_close(fft.ifftshift(y), x)
-        x = xp.asarray([0, 1, 2, 3, 4, -5, -4, -3, -2, -1])
-        y = xp.asarray([-5, -4, -3, -2, -1, 0, 1, 2, 3, 4])
+        x = xp.asarray([0., 1, 2, 3, 4, -5, -4, -3, -2, -1])
+        y = xp.asarray([-5., -4, -3, -2, -1, 0, 1, 2, 3, 4])
         xp_assert_close(fft.fftshift(x), y)
         xp_assert_close(fft.ifftshift(y), x)
 
@@ -365,8 +365,8 @@ class TestFFTShift:
     @skip_if_array_api_backend('torch')
     @array_api_compatible
     def test_axes_keyword(self, xp):
-        freqs = xp.asarray([[0, 1, 2], [3, 4, -4], [-3, -2, -1]])
-        shifted = xp.asarray([[-1, -3, -2], [2, 0, 1], [-4, 3, 4]])
+        freqs = xp.asarray([[0., 1, 2], [3, 4, -4], [-3, -2, -1]])
+        shifted = xp.asarray([[-1., -3, -2], [2, 0, 1], [-4, 3, 4]])
         xp_assert_close(fft.fftshift(freqs, axes=(0, 1)), shifted)
         xp_assert_close(fft.fftshift(freqs, axes=0), fft.fftshift(freqs, axes=(0,)))
         xp_assert_close(fft.ifftshift(shifted, axes=(0, 1)), freqs)
@@ -384,14 +384,14 @@ class TestFFTShift:
             [0, 1],
             [2, 3],
             [4, 5]
-        ])
+        ], dtype=xp.float64)
 
         # shift in dimension 0
         shift_dim0 = xp.asarray([
             [4, 5],
             [0, 1],
             [2, 3]
-        ])
+        ], dtype=xp.float64)
         xp_assert_close(fft.fftshift(freqs, axes=0), shift_dim0)
         xp_assert_close(fft.ifftshift(shift_dim0, axes=0), freqs)
         xp_assert_close(fft.fftshift(freqs, axes=(0,)), shift_dim0)
@@ -402,7 +402,7 @@ class TestFFTShift:
             [1, 0],
             [3, 2],
             [5, 4]
-        ])
+        ], dtype=xp.float64)
         xp_assert_close(fft.fftshift(freqs, axes=1), shift_dim1)
         xp_assert_close(fft.ifftshift(shift_dim1, axes=1), freqs)
 
@@ -411,7 +411,7 @@ class TestFFTShift:
             [5, 4],
             [1, 0],
             [3, 2]
-        ])
+        ], dtype=xp.float64)
         xp_assert_close(fft.fftshift(freqs, axes=(0, 1)), shift_dim_both)
         xp_assert_close(fft.ifftshift(shift_dim_both, axes=(0, 1)), freqs)
         xp_assert_close(fft.fftshift(freqs, axes=[0, 1]), shift_dim_both)

--- a/scipy/special/tests/test_support_alternative_backends.py
+++ b/scipy/special/tests/test_support_alternative_backends.py
@@ -8,11 +8,18 @@ from scipy.conftest import array_api_compatible
 from scipy import special
 from scipy._lib._array_api import xp_assert_close
 from scipy._lib.array_api_compat import numpy as np
-import numpy.array_api as np_array_api
+
+try:
+    import array_api_strict
+    HAVE_ARRAY_API_STRICT = True
+except ImportError:
+    HAVE_ARRAY_API_STRICT = False
 
 
+@pytest.mark.skipif(not HAVE_ARRAY_API_STRICT,
+                    reason="`array_api_strict` not installed")
 def test_dispatch_to_unrecognize_library():
-    xp = np_array_api
+    xp = array_api_strict
     f = get_array_special_func('ndtr', xp=xp, n_array_args=1)
     x = [1, 2, 3]
     res = f(xp.asarray(x))

